src/ir.rs

```diff
@@ -709,6 +720,29 @@ pub enum DomainGoal {
     /// like `Box<T>`, it is true if `T` is local.
     IsLocal(Ty),
 
+    /// True if a type is *not* considered to have been "defined" by the current crate. This is
+    /// false for a `struct Foo { }` but true for a `extern struct Foo { }`. However, for
+    /// fundamental types like `Box<T>`, it is true if `T` is external.
+    IsExternal(Ty),
+
+    /// True if a type both external and its type parameters are recursively external
+    ///
+    /// More formally, for each non-fundamental struct S<P0..Pn> that is external:
+    /// forall<P0..Pn> {
+    ///     IsDeeplyExternal(S<P0...Pn>) :-
+    ///         IsDeeplyExternal(P0),
+    ///         ...
+    ///         IsDeeplyExternal(Pn)
+    /// }
+    ///
+    /// For each fundamental struct P<P0>,
+    ///
+    /// forall<P0> { IsDeeplyExternal(S<P0>) :- IsDeeplyExternal(P0) }
```

Comment:
According to these comments, it seems to me that there is no difference between `IsDeeplyExternal` for non-fundamental-extern type and for fundamental-extern types.

Reply:
I think the right way to think about it is that structs are either local, extern, or fundamental, but not (e.g.) "extern fundamental".